src/extensions/compression/deflate.rs

```diff
@@ -0,0 +1,289 @@
+                        }
+                        seen_client_max_window_bits = true;
+                    }
+
+                    // Offer with unknown parameter MUST be declined.
+                    _ => {
+                        continue 'outer;
+                    }
+                }
+            }
+
+            return Some((to_header_value(&agreed), DeflateContext::new(Role::Server, config)));
+        }
+
+        None
+    }
+
+    pub(crate) fn accept_response(&self, agreed: &[Param]) -> Result<DeflateContext, DeflateError> {
+        let mut config =
+            DeflateConfig { compression: self.compression, ..DeflateConfig::default() };
+        for param in agreed {
+            match param.name() {
+                SERVER_NO_CONTEXT_TAKEOVER => {
+                    config.server_no_context_takeover = true;
+                }
+
+                CLIENT_NO_CONTEXT_TAKEOVER => {
+                    config.client_no_context_takeover = true;
+                }
+
+                SERVER_MAX_WINDOW_BITS => {}
+                CLIENT_MAX_WINDOW_BITS => {}
+
+                _ => {
+                    //
+                }
+            }
+        }
+        Ok(DeflateContext::new(Role::Client, config))
+    }
+}
+
+#[derive(Debug)]
+/// Manages per message compression using DEFLATE.
+pub struct DeflateContext {
+    role: Role,
+    config: DeflateConfig,
+    compressor: Compress,
+    decompressor: Decompress,
+}
+
+impl DeflateContext {
+    fn new(role: Role, config: DeflateConfig) -> Self {
+        DeflateContext {
+            role,
+            config,
+            compressor: Compress::new(config.compression, false),
+            decompressor: Decompress::new(false),
+        }
+    }
+
+    fn own_context_takeover(&self) -> bool {
+        match self.role {
+            Role::Server => !self.config.server_no_context_takeover,
+            Role::Client => !self.config.client_no_context_takeover,
+        }
+    }
+
+    fn peer_context_takeover(&self) -> bool {
+        match self.role {
+            Role::Server => !self.config.client_no_context_takeover,
+            Role::Client => !self.config.server_no_context_takeover,
+        }
+    }
+
+    // Compress the data of message.
+    pub(crate) fn compress(&mut self, data: &[u8]) -> Result<Vec<u8>, DeflateError> {
+        // https://datatracker.ietf.org/doc/html/rfc7692#section-7.2.1
+        // 1. Compress all the octets of the payload of the message using DEFLATE.
+        let mut output = Vec::with_capacity(data.len());
+        let before_in = self.compressor.total_in() as usize;
+        while (self.compressor.total_in() as usize) - before_in < data.len() {
+            let offset = (self.compressor.total_in() as usize) - before_in;
+            match self
+                .compressor
+                .compress_vec(&data[offset..], &mut output, FlushCompress::None)
+                .map_err(|e| DeflateError::Compress(e.into()))?
```

Comment:
NB: What if we use the `#[from]` (from `thiserror`) on the error structure? Would it help to just use `?` after `compress_vec`? (e.g. spares us an explicit conversion with `map_err`)

Reply:
`compress_vec(...)?` only works if `DeflateError::Compress(flate2::CompressError)`.

I was thinking `Compress(std::io::Error)` and `Decompress(std::io::Error)` are better if we decide to support more PMCEs, and make them generic. Explicit conversions are required because both variants are `std::io::Error` (`#[from]` can't be used on both).